File: output/cpp/tests/mod_b.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

class S2FrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    S2FrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class S2FrameContext {
public:
    S2FrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    S2FrameContext(S2FrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class S2Compartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<S2FrameEvent> forward_event;
    std::unique_ptr<S2Compartment> parent_compartment;

    explicit S2Compartment(const std::string& state) : state(state) {}

    std::unique_ptr<S2Compartment> clone() const {
        auto c = std::make_unique<S2Compartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class S2 {
private:
    std::vector<std::unique_ptr<S2Compartment>> _state_stack;
    std::unique_ptr<S2Compartment> __compartment;
    std::unique_ptr<S2Compartment> __next_compartment;
    std::vector<S2FrameContext> _context_stack;

    void __kernel(S2FrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            S2FrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                S2FrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    S2FrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(S2FrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "A") {
            _state_A(__e);
        }
    }

    void __transition(std::unique_ptr<S2Compartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_A(S2FrameEvent& __e) {
        if (__e._message == "e") {
            auto __new_compartment = std::make_unique<S2Compartment>("A");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

public:
    S2() {
        __compartment = std::make_unique<S2Compartment>("A");
        S2FrameEvent __frame_event("$>");
        S2FrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void e() {
        S2FrameEvent __e("e");
        S2FrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

// Stub functions for placeholder calls
void native() {}
void x() {}

// TAP test harness
int main() {
    printf("TAP version 14\n");
    printf("1..1\n");
    try {
        S2 s;
        s.e();
        printf("ok 1 - mod_b\n");
    } catch (...) {
        printf("not ok 1 - mod_b\n");
    }
    return 0;
}

File: output/go/tests/mod_b.go

package main

import "fmt"

type S2FrameEvent struct {
    _message    string
    _parameters map[string]any
}

type S2FrameContext struct {
    _event  S2FrameEvent
    _return any
    _data   map[string]any
}

type S2Compartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *S2FrameEvent
    parentCompartment *S2Compartment
}

func newS2Compartment(state string) *S2Compartment {
    return &S2Compartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *S2Compartment) copy() *S2Compartment {
    nc := &S2Compartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type S2 struct {
    _state_stack []*S2Compartment
    __compartment *S2Compartment
    __next_compartment *S2Compartment
    _context_stack []S2FrameContext
}

func NewS2() *S2 {
    s := &S2{}
    s._state_stack = make([]*S2Compartment, 0)
    s._context_stack = make([]S2FrameContext, 0)
    s.__compartment = newS2Compartment("A")
    s.__next_compartment = nil
    __frame_event := S2FrameEvent{_message: "$>", _parameters: nil}
    __ctx := S2FrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *S2) __kernel(__e *S2FrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &S2FrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &S2FrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &S2FrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *S2) __router(__e *S2FrameEvent) {
    switch s.__compartment.state {
    case "A":
        s._state_A(__e)
    }
}

func (s *S2) __transition(next *S2Compartment) {
    s.__next_compartment = next
}

func (s *S2) E() {
    __e := S2FrameEvent{_message: "E"}
    __ctx := S2FrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *S2) _state_A(__e *S2FrameEvent) {
    if __e._message == "E" {
        __compartment := newS2Compartment("A")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func main() {
	fmt.Println("TAP version 14")
	fmt.Println("1..1")
	s := NewS2()
	s.E()
	fmt.Println("ok 1 - mod_b")
}
